src/resolutions.js

```diff
@@ -0,0 +1,101 @@
+/* @flow */
+import path from 'path';
+import semver from 'semver';
+import minimatch from 'minimatch';
+import map from './util/map.js';
+import type Config from './config.js';
+import type {Reporter} from './reporters/index.js';
+import PackageRequest from './package-request';
+
+const DIRECTORY_SEPARATOR = '/';
+
+export type Resolution = {
+  name: string,
+  version: string,
+  pattern: string,
+};
+
+export type ResolutionMap = {
+  [packageName: string]: Array<Resolution>,
+};
+
+export type ResolvedPatternsMap = {
+  [reqPattern: string]: string,
+};
+
+export type ResolutionEntry = {
+  [packageName: string]: string,
+};
+
+export default class Resolutions {
+  constructor(config: Config) {
+    this.resolutionsByPackage = map();
+    this.resolvedPatterns = map();
+    this.config = config;
+    this.reporter = config.reporter;
+  }
+
+  resolutionsByPackage: ResolutionMap;
+  resolvedPatterns: ResolvedPatternsMap;
+  config: Config;
+  reporter: Reporter;
+
+  init(resolutions: ?ResolutionEntry = {}) {
+    Object.keys(resolutions).map(pattern => {
+      const info = this.parsePatternInfo(pattern, resolutions[pattern]);
+
+      if (info) {
+        const resolution = this.resolutionsByPackage[info.name] || [];
+        this.resolutionsByPackage[info.name] = [...resolution, info];
+      }
+    });
+  }
+
+  getResolutionByPattern(pattern: string): ?string {
+    return this.resolvedPatterns[pattern];
+  }
+
+  parsePatternInfo(pattern: string, version: string): ?Object {
+    const directories = pattern.split(DIRECTORY_SEPARATOR);
+    const name = directories.pop();
+
+    if (!name) {
+      this.reporter.warn(this.reporter.lang('invalidResolutionName', pattern));
+      return null;
+    }
+
+    return {
+      name,
+      version,
+      pattern,
+    };
+  }
+
+  find(reqPattern: string, parentNames: Array<string>): ?string {
+    const {name, range} = PackageRequest.normalizePattern(reqPattern);
+    const resolutions = this.resolutionsByPackage[name];
+    let resolvedPattern = null;
+
+    if (!resolutions) {
+      return resolvedPattern;
+    }
+
+    const {version} = resolutions.find(({pattern}) => minimatch(path.join(...parentNames, name), pattern)) || {};
+
+    if (version) {
+      resolvedPattern = `${name}@${version}`;
+
+      if (!semver.valid(version)) {
+        this.reporter.warn(this.reporter.lang('invalidResolutionVersion', resolvedPattern));
+      }
```

Comment:
I think it would be best if we could support it for 1.0, yep. But we can merge this PR before that I you prefer, and iterate on it.